lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,778 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Data structures and encoding for `offer` messages.
+//!
+//! An [`Offer`] is built by the merchant for the user-pays-merchants flow or may be parsed from a
```

Comment:
The "user-pays-merchant" terminology is from the BOLT and is used to differentiate from "merchant-pays-user" (e.g., refunds, ATM) flow, which I use in a future commit. I could use "recipient" instead of the first use of "merchant" here to make it more generic. But not sure if I want to deviate from the spec by adding a new "flow". Open to other ideas so long as we can make it work with the `invoice_request` docs (see #1597 for current draft), which will be expanded with another builder for the "merchant-pays-user" flow.

Reply:
Maybe we could do `payer-offers-money` vs `payer-pays-recipient` (maybe the word "offer" is confusing there)? ATM isn't really captured by "merchant" either.. I think we've historically not cared too much about sticking with spec naming/copy if there's a reason to deviate, like with `wumbo`, `scid_privacy` vs `alias`, etc.

Reply:
From my PoV, I'm fine with the BOLT copy because it's protocol-dev-facing and gets the point across. For wallet-dev-facing documentation, though, IMO there are different requirements and it'd be best to be inclusive of the range of use cases. 

Given this, maybe we could have an expanded `Use Cases` section in the top-level docs, which include the user-pays-merchant/ATM/etc, and `Offer` could be documented as "an offer to be paid" and `InvoiceRequest` as "an offer of money", rather than each referencing a specific "flow"?

Reply:
> From my PoV, I'm fine with the BOLT copy because it's protocol-dev-facing and gets the point across. For wallet-dev-facing documentation, though, IMO there are different requirements and it'd be best to be inclusive of the range of use cases.

Could you enumerate these use cases?

> 
> Given this, maybe we could have an expanded `Use Cases` section in the top-level docs, which include the user-pays-merchant/ATM/etc, and `Offer` could be documented as "an offer to be paid" and `InvoiceRequest` as "an offer of money", rather than each referencing a specific "flow"?

Yeah, I agree that the `offers` top-level module should have an expanded explanation of use cases and how the various types relate to them. I'd prefer to wait until all the types exist before writing this.

In the `offer` submodule, I'd like to stick to showing basic example code, though, for the specific types involved. I've updated the wording a bit. Let me know what you think.